[FREELDR:PC98] Don't do the HighRes check everytime a character is written with writechr.

HBelusca opened the pull request:
## Purpose

See this PR's title. I suppose that this setting won't change between calls (when e.g. done from the other `writestr` helpers).

## Proposed changes

Do the HighRes check and cache the computed value of `VramSegment` only once, then reuse that latter one.

## TODO

- [x] Retest again once other blocker PR #4358 gets fixed.

HBelusca commented:
Cc @disean.

disean commented on boot/freeldr/freeldr/arch/realmode/helpers_pc98.inc:
```diff
@@ -80,15 +80,22 @@ writechr:
+    mov word ptr VramSegment, ax
+
+.writechr_test_done2:
+    push ax
     pop es
```
`mov es, ax`?

disean commented on boot/freeldr/freeldr/arch/realmode/helpers_pc98.inc:
```diff
@@ -80,15 +80,22 @@ writechr:
     pushf
     pusha
 
+    mov ax, word ptr VramSegment
```
There is a nice `LES` instruction you can use. This suggestion has been tested to work correctly:
```writechr:
    pushf
    pusha

    /* Check if the VRAM segment was initialized */
    mov dx, word ptr ds:[VramSegment]
    test dx, dx
    jnz .writechr_write

    /* High-resolution mode check */
    test byte ptr ds:[HEX(501)], HEX(08)
    mov dx, HEX(0A000)
    jz .writechr_test_done
    mov dh, HEX(0E)

.writechr_test_done:
    mov word ptr ds:[VramSegment], dx

.writechr_write:
    /* Load the VRAM segment and offset */
    les di, dword ptr ds:[VramOffset]

    /* Two bytes per character */
    xor ah, ah

    /* Write ASCII directly to the VRAM */
    stosw

    /* Update the start position */
    mov word ptr ds:[VramOffset], di

    popa
    popf
    ret

VramOffset:
    .word 0
VramSegment:
    .word 0

HBelusca replied:
I've updated the code accordingly, please review :)

disean commented:
Come to think of it, I wrote the hi-reso video support code with little knowledge of the PC-98 architecture, and I just realized that hi-reso models don't meet the minimum hardware requirements for ReactOS (https://www.pc-9800.net/db_98/dbh98_index.htm). ROS PC-98 targets PC-9821 hardware with PCI bus. I suggest to remove the VRAM check at all.

binarymaster commented:
They all are non-Pentium? Machines without PCI should be possible (even in PC/AT case), though I don't rememeber if our Standard HAL could handle these properly.

disean commented:
Yeah, they use processors older than Pentium.

HBelusca commented:
Hi @disean and @binarymaster , I've updated the PR, could you please retest it? (I've also updated the bootsector with some of the optimizations @disean suggested.)

disean commented on boot/freeldr/freeldr/arch/realmode/helpers_pc98.inc:
```diff
@@ -80,27 +80,39 @@ writechr:
     popf
     ret
 
+VramSegOff:
 VramOffset:
```
Why is there two symbols in the first place?

HBelusca replied:
The redundant `VramSegOff` is just to understand what's going to be read by the `les di, dword ptr ds:[VramSegOff]` instruction (mentioning only "VramOffset" in it instead could look weird at first).

HBelusca commented on boot/freeldr/freeldr/arch/realmode/helpers_pc98.inc:
```diff
@@ -80,27 +80,39 @@ writechr:
-
+    mov dx, HEX(0A000) // Suppose normal mode
+    jz short .writechr_test_done
+    mov dh, HEX(E0)    // Change 0xA000 to 0xE000, use hi-res mode
```
@disean Note that here I fixed the value to be E0, instead of 0E as in your proposal, because it had to change the A0 to E0.
